src/WebForms/SiteMapProvider.cs

```diff
@@ -0,0 +1,473 @@
+    {
+        get
+        {
+            return _parentProvider;
+        }
+        set
+        {
+            _parentProvider = value;
+        }
+    }
+
+    public string ResourceKey
+    {
+        get
+        {
+            return _resourceKey;
+        }
+        set
+        {
+            _resourceKey = value;
+        }
+    }
+
+    public virtual SiteMapProvider RootProvider
+    {
+        get
+        {
+            if (_rootProvider == null)
+            {
+                lock (_lock)
+                {
+                    if (_rootProvider == null)
+                    {
+                        Hashtable providers = new Hashtable();
+                        SiteMapProvider candidate = this;
+
+                        providers.Add(candidate, null);
+                        while (candidate.ParentProvider != null)
+                        {
+                            if (providers.Contains(candidate.ParentProvider))
+                                throw new ProviderException(SR.GetString(SR.SiteMapProvider_Circular_Provider));
+
+                            candidate = candidate.ParentProvider;
+                            providers.Add(candidate, null);
+                        }
+
+                        _rootProvider = candidate;
+                    }
+                }
+            }
+
+            return _rootProvider;
+        }
+    }
+
+    public virtual SiteMapNode RootNode
+    {
+        get
+        {
+            SiteMapNode node = GetRootNodeCore();
+            return ReturnNodeIfAccessible(node);
+        }
+    }
+
+    public bool SecurityTrimmingEnabled
+    {
+        get
+        {
+            return _securityTrimmingEnabled;
+        }
+    }
+
+    public event SiteMapResolveEventHandler SiteMapResolve;
+
+    /// <devdoc>
+    ///    <para>Add single node to provider.</para>
+    /// </devdoc>
+    protected virtual void AddNode(SiteMapNode node)
+    {
+        AddNode(node, null);
+    }
+
+    protected internal virtual void AddNode(SiteMapNode node, SiteMapNode parentNode)
+    {
+        throw new NotImplementedException();
+    }
+
+    public virtual SiteMapNode FindSiteMapNode(HttpContext context)
+    {
+        if (context == null)
+        {
+            return null;
+        }
+
+        string rawUrl = context.Request.RawUrl;
+
+        SiteMapNode result = null;
+
+        // First check the RawUrl
+        result = FindSiteMapNode(rawUrl);
+
+        if (result == null)
+        {
+            int queryStringIndex = rawUrl.IndexOf("?", StringComparison.Ordinal);
+            if (queryStringIndex != -1)
+            {
+                // check the RawUrl without querystring
+                result = FindSiteMapNode(rawUrl.Substring(0, queryStringIndex));
+            }
+
+            if (result == null)
+            {
+                Page page = context.GetRequiredFeature<IHttpHandlerFeature>().Current as Page;
+
+                //TODO check with @taylor
```

Comment:
@twsouthwick take a look

Reply:
Can you add an extension method for `Page? GetCurrentPage(this HttpContext context)`?